homeassistant/components/nextbus/config_flow.py

```diff
@@ -0,0 +1,202 @@
+        if agency not in agency_tags:
+            return self.async_abort(reason="invalid_agency")
+
+        route_tags = await self.hass.async_add_executor_job(
+            _get_route_tags, self._client, agency
+        )
+        if route not in route_tags:
+            return self.async_abort(reason="invalid_route")
+
+        stop_tags = await self.hass.async_add_executor_job(
+            _get_stop_tags, self._client, agency, route
+        )
+        if stop not in stop_tags:
+            return self.async_abort(reason="invalid_stop")
+
+        data = {
+            CONF_AGENCY: agency,
+            CONF_ROUTE: route,
+            CONF_STOP: stop,
+        }
+
+        # Abort if duplicate entries exist
+        self._async_abort_entries_match(self.data)
+
+        return self.async_create_entry(
+            title=(
+                config_input.get(CONF_NAME)
+                or f"{config_input[CONF_AGENCY]} {config_input[CONF_ROUTE]}"
+            ),
+            data=data,
+        )
+
+    async def async_step_user(
+        self,
+        user_input: dict[str, str] | None = None,
+    ) -> FlowResult:
+        """Handle a flow initiated by the user."""
+        return await self.async_step_agency(user_input)
+
+    async def async_step_agency(
+        self,
+        user_input: dict[str, str] | None = None,
+    ) -> FlowResult:
+        """Select agency."""
+        if user_input is not None:
+            self.data[CONF_AGENCY] = user_input[CONF_AGENCY]
+
+            return await self.async_step_route()
+
+        self._agency_tags = await self.hass.async_add_executor_job(
+            _get_agency_tags, self._client
+        )
+
+        return self.async_show_form(
+            step_id="agency",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_AGENCY): _dict_to_select_selector(
+                        self._agency_tags
+                    ),
+                }
+            ),
+        )
+
+    async def async_step_route(
+        self,
+        user_input: dict[str, str] | None = None,
+    ) -> FlowResult:
+        """Select route."""
+        if user_input is not None:
+            self.data[CONF_ROUTE] = user_input[CONF_ROUTE]
+
+            return await self.async_step_stop()
+
+        self._route_tags = await self.hass.async_add_executor_job(
+            _get_route_tags, self._client, self.data[CONF_AGENCY]
+        )
+
+        return self.async_show_form(
+            step_id="route",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_ROUTE): _dict_to_select_selector(
+                        self._route_tags
+                    ),
+                }
+            ),
+        )
+
+    async def async_step_stop(
+        self,
+        user_input: dict[str, str] | None = None,
+    ) -> FlowResult:
+        """Select stop."""
+
+        if user_input is not None:
+            self.data[CONF_STOP] = user_input[CONF_STOP]
```

Comment:
For readability, I left the similar assignments grouped together. Combining these to one line will not have a performance impact yet makes it less clear how the tag values are used.

I could get rid of these intermediate variables entirely as a minor optimization, but I left them so that if a KeyError is raised, it's apparent whether or not it was with `self.data` or with `self._*_tags`.